src/main/scala/com/evolution/scache/LoadingCache.scala

```diff
@@ -116,50 +119,114 @@ private[scache] object LoadingCache {
-                                                  deferred
-                                                    .getOrError
-                                                    .map { entry =>
-                                                      entry
-                                                        .value
-                                                        .asRight[F[V]]
-                                                        .asRight[A]
+                                                  entry
+                                                    .release1
+                                                    .start
+                                                    .flatMap { _ =>
+                                                      entryRef
+                                                        .get
+                                                        .map {
+                                                          case state: EntryState.Value[F, V] =>
+                                                            state
+                                                              .entry
+                                                              .value
+                                                              .asRight[F[V]]
+                                                              .asRight[A]
+                                                          case state: EntryState.Loading[F, V] =>
+                                                            state
+                                                              .deferred
+                                                              .getOrError
+                                                              .map(_.value)
+                                                              .asLeft[V]
+                                                              .asRight[A]
+                                                          case EntryState.Removed =>
+                                                            a.asLeft[Either[F[V], V]]
+                                                        }
                                                     }
                                               }
-                                              .map { _.asRight[Int] }
-                                          case false =>
-                                            (counter + 1)
-                                              .asLeft[Either[A, Either[F[V], V]]]
-                                              .pure[F]
+
+                                          // Entry contains a _different_ loading value
+                                          // (for example as a result of concurrent `remove` followed by another `getOrUpdate1`),
+                                          // so we return new value computation and release our value immediately.
+                                          case (state: EntryState.Loading[F, V], _) =>
```

Comment:
Basically in all cases `case (state: EntryState.Loading[F, V], set) if state.deferred == deferred =>`
we need to `release`, hence I'd suggest to refactor a bit:


```scala
case state => 

entry
                                              .release1
                                              .start
                                              .as {
                                              state match {
                                              }
                                              }

```

Reply:
agreed, done